modules/contact/src/userobjects/WeightedGapUserObject.C

```diff
@@ -0,0 +1,155 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "WeightedGapUserObject.h"
+#include "MooseVariableField.h"
+#include "SubProblem.h"
+#include "MortarUtils.h"
+#include "MooseUtils.h"
+#include "MortarContactUtils.h"
+#include "libmesh/quadrature.h"
+
+InputParameters
+WeightedGapUserObject::validParams()
+{
+  InputParameters params = MortarUserObject::validParams();
+  params += MortarConsumerInterface::validParams();
+  params += TwoMaterialPropertyInterface::validParams();
+  params.addRequiredCoupledVar("disp_x", "The x displacement variable");
+  params.addRequiredCoupledVar("disp_y", "The y displacement variable");
+  params.addCoupledVar("disp_z", "The z displacement variable");
+  params.set<bool>("use_displaced_mesh") = true;
+  params.set<bool>("interpolate_normals") = false;
+  params.set<ExecFlagEnum>("execute_on") = {EXEC_LINEAR, EXEC_NONLINEAR};
+  params.suppressParameter<ExecFlagEnum>("execute_on");
+  return params;
+}
+
+WeightedGapUserObject::WeightedGapUserObject(const InputParameters & parameters)
+  : MortarUserObject(parameters),
```

Comment:
General comment. We have now another layer of mortar objects which draw on base class functionality. Stuff like `correct_edge_dropping` and `minimum_projection_angle` (or was it maximum?) will need to be synchronize among these uos and constraints. Maybe a good opportunity to add more functionality to the action and not encourage the user to set things up by themselves.

Reply:
Okay, gather all these things into a small issue. At least, I won't forget: https://github.com/idaholab/moose/issues/23889